### PR TITLE
resolve #335: git ship: retain author

### LINKED ISSUE #335: git ship: retain author
Set the --author flag on the squash commit to be whatever author has the most commits in the shipped branch. This is for when someone other than the author ships the branch.

### PR DESCRIPTION
**Pull request by charlierudolph**
@kevgo @allewun @alexdavid 

git ship sets the author to whoever has the most commits in the shipped branch

### CODE REVIEW

File: features/git-ship/current_branch/on_coworker_feature_branch.feature [hunk 1]
```diff
@@ -0,0 +1,35 @@
+Feature: git ship: shipping a coworkers branch
+
+  As a developer shipping a coworker's feature branch
+  I want my coworker to be the author of the commit added to the main branch
+  So my coworker is given credit for their work
```
Comment (kevgo) [inline features/git-ship/current_branch/on_coworker_feature_branch.feature]: So **that** my coworker...

File: features/git-ship/current_branch/on_feature_branch_without_open_changes.feature [hunk 2]
```diff
@@ -38,8 +38,10 @@ Feature: git ship: shipping the current feature branch
   Scenario: feature branch with non-pulled updates in the repo
     Given I have a feature branch named "feature"
     And the following commit exists in my repository
-      | BRANCH  | LOCATION | FILE NAME    | FILE CONTENT    |
-      | feature | remote   | feature_file | feature content |
+      | BRANCH  | LOCATION         | FILE NAME    | FILE CONTENT          |
```
Comment (kevgo) [inline features/git-ship/current_branch/on_feature_branch_without_open_changes.feature]: These changes seem unrelated to this PR. Can you `git reset master` and commit them to another branch?

File: features/support/commit_helpers.rb [hunk 3]
```diff
@@ -76,13 +76,16 @@ def committed_files sha
...
+def commits_for_branch branch_name, keys
+  array_output_of("git log #{branch_name} --format='%h|%s|%ae'").map do |commit|
+    sha, message, author = commit.split('|')
+    next if message == 'Initial commit'
+    {
+      author: author,
+      branch: branch_name,
+      message: message,
+      file_name: committed_files(sha)
+    }.select { |k, _| keys.include?(k) }
```
Comment (kevgo) [inline features/support/commit_helpers.rb]: Rename `k` to `key`

Comment (kevgo) [inline features/support/commit_helpers.rb]: Remove unnecessary parentheses around `k`

File: features/support/env.rb [hunk 4]
```diff
@@ -19,17 +19,13 @@
   # Create the local repository
   clone_repository remote_repository_path, local_repository_path
   at_path local_repository_path do
+    configure_git 'local'
```
Comment (kevgo) [inline features/support/env.rb]: 'local' is a strange user name

File: features/git-ship/current_branch/on_coworker_feature_branch.feature [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+      | feature | git merge --no-edit main                                                |
+      | feature | git checkout main                                                       |
+      | main    | git merge --squash feature                                              |
+      | main    | git commit --author="coworker <coworker@example.com>" -m 'feature done' |
+      | main    | git push                                                                |
+      | main    | git push origin :feature                                                |
+      | main    | git branch -D feature                                                   |
+    And I have the following commits
+      | BRANCH | LOCATION         | MESSAGE      | FILE NAME     | AUTHOR               |
+      | main   | local and remote | feature done | coworker_file | coworker@example.com |
```
Comment (kevgo) [inline features/git-ship/current_branch/on_coworker_feature_branch.feature]: An edge case to consider here is when two people have the same number of commits. In this case it should pick the person that did an earlier commit as the author.

File: features/git-ship/current_branch/on_feature_branch_without_open_changes.feature (hunk 2 above)
Comment (charlierudolph) [inline features/git-ship/current_branch/on_feature_branch_without_open_changes.feature]: These are related. Without these additional commits, its the coworker who is the only with a commit on the branch. This is because we implement remote commits as being a coworker committing them instead of the generic user committing them from another machine

Comment (kevgo) [inline features/git-ship/current_branch/on_feature_branch_without_open_changes.feature]: Ha, that's why I implemented remote commits as from the normal user initially (okay, with some `git reset` madness thrown in). Then we changed it to a more clever hack that uses the co-worker repo. And now we need more hacks to work around the first hack. I think we should clean this up. Having specs behave unintuitively if I use remote commits, and forcing me to create 3 commits in that case, is not good style.

Comment (charlierudolph) [inline features/git-ship/current_branch/on_feature_branch_without_open_changes.feature]: Okay I will do a separate PR implementing remote commits as another git repo for the local user. Will also try to lazily create repos as we need them

File: features/git-ship/current_branch/on_coworker_feature_branch.feature (hunk 1 above)
Comment (charlierudolph) [inline features/git-ship/current_branch/on_coworker_feature_branch.feature]: done

File: features/support/commit_helpers.rb (hunk 3 above)
Comment (charlierudolph) [inline features/support/commit_helpers.rb]: done

File: features/support/env.rb (hunk 4 above)
Comment (charlierudolph) [inline features/support/env.rb]: renamed to developer in #343

File: features/git-ship/current_branch/on_coworker_feature_branch.feature (hunk 5 above)
Comment (charlierudolph) [inline features/git-ship/current_branch/on_coworker_feature_branch.feature]: Can this be a separate PR? It will make the logic for determining the author much more complicated.

Comment (kevgo) [inline features/git-ship/current_branch/on_coworker_feature_branch.feature]: Absolutely

**Comment by charlierudolph** [comment]
all comments addressed

**Comment by kevgo** [comment]
:ship:
